crates/base/src/vector/veci8.rs

```diff
@@ -0,0 +1,338 @@
+        if dims == 0 || dims as usize != data.len() {
+            return None;
+        }
+        let (sum, l2_norm) = i8_precompute(&data, alpha, offset);
+        Some(unsafe { Self::new_unchecked(dims, data, alpha, offset, sum, l2_norm) })
+    }
+
+    /// # Safety
+    ///
+    /// * `dims` must be in `1..=65535`.
+    /// * `dims` must be equal to `values.len()`.
+    #[inline(always)]
+    pub unsafe fn new_unchecked(
+        dims: u16,
+        data: Vec<I8>,
+        alpha: F32,
+        offset: F32,
+        sum: F32,
+        l2_norm: F32,
+    ) -> Self {
+        Veci8Owned {
+            dims,
+            data,
+            alpha,
+            offset,
+            sum,
+            l2_norm,
+        }
+    }
+
+    pub fn data(&self) -> &[I8] {
+        &self.data
+    }
+
+    pub fn alpha(&self) -> F32 {
+        self.alpha
+    }
+
+    pub fn alpha_mut(&mut self) -> &mut F32 {
+        &mut self.alpha
+    }
+
+    pub fn offset_mut(&mut self) -> &mut F32 {
+        &mut self.offset
+    }
+
+    pub fn offset(&self) -> F32 {
+        self.offset
+    }
+
+    pub fn sum(&self) -> F32 {
+        self.sum
+    }
+
+    pub fn l2_norm(&self) -> F32 {
+        self.l2_norm
+    }
+
+    pub fn dims(&self) -> u16 {
+        self.dims
+    }
+}
+
+impl VectorOwned for Veci8Owned {
+    // For i8 quantization, the scalar type is used for type in kmeans, it use F32 to store the centroids and examples in the kmeans.
+    type Scalar = F32;
+    type Borrowed<'a> = Veci8Borrowed<'a>;
+
+    #[inline(always)]
+    fn dims(&self) -> u16 {
+        self.dims
+    }
+
+    fn for_borrow(&self) -> Veci8Borrowed<'_> {
+        Veci8Borrowed {
+            dims: self.dims,
+            data: &self.data,
+            alpha: self.alpha,
+            offset: self.offset,
+            sum: self.sum,
+            l2_norm: self.l2_norm,
+        }
+    }
+
+    fn to_vec(&self) -> Vec<F32> {
+        i8_dequantization(&self.data, self.alpha, self.offset)
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct Veci8Borrowed<'a> {
+    dims: u16,
+    data: &'a [I8],
+    alpha: F32,
+    offset: F32,
+    // sum of a_i * alpha, precomputed for dot
+    sum: F32,
+    // l2 norm of original f_i, precomputed for l2
+    l2_norm: F32,
+}
+
+impl<'a> Veci8Borrowed<'a> {
+    #[inline(always)]
+    pub fn new(dims: u16, data: &'a [I8], alpha: F32, offset: F32) -> Veci8Borrowed<'a> {
+        Self::new_checked(dims, data, alpha, offset).unwrap()
+    }
+
+    #[inline(always)]
+    pub fn new_checked(dims: u16, data: &'a [I8], alpha: F32, offset: F32) -> Option<Self> {
+        if dims == 0 || dims as usize != data.len() {
+            return None;
+        }
+        let (sum, l2_norm) = i8_precompute(data, alpha, offset);
+        Some(unsafe { Self::new_unchecked(dims, data, alpha, offset, sum, l2_norm) })
+    }
+
+    #[inline(always)]
+    pub fn new_checked_without_precomputed(
```

Comment:
We never need this function. PostgreSQL may take advantage of bitwise equality so we should guarantee that there is only one memory representation for same vectors.